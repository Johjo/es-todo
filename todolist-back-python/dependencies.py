from dataclasses import dataclass, field, replace
from enum import Enum
from typing import Any

class ResourceType(str, Enum):
    use_case = "use_case"
    adapter = "adapter"
    path = "path"
    infrastructure = "infrastructure"


@dataclass(frozen=True)
class Dependencies:
    factory: dict[Any, Any] = field(default_factory=dict)

    def feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return self._feed(resource_type=resource_type, resource=resource, factory=factory)

    def feed_use_case(self, use_case: Any, use_case_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.use_case, resource=use_case, factory=use_case_factory)

    def feed_adapter(self, port: Any, adapter_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.adapter, resource=port, factory=adapter_factory)

    def feed_path(self, path: str, factory) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.path, resource=path, factory=factory)

    def feed_infrastructure(self, infrastructure: Any, factory):
        return self._feed(resource_type=ResourceType.infrastructure, resource=infrastructure, factory=factory)

    def _feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return replace(self, factory={**self.factory, (resource_type, resource): factory})

    def get_use_case(self, use_case: Any) -> Any:
        return self._get_resource(resource_type=ResourceType.use_case, resource=use_case)

    def get_query(self, query: Any) -> Any:
        return self.get_use_case(query)

    def get_infrastructure(self, infrastructure) -> Any:
        resource = self._get_resource(resource_type=ResourceType.infrastructure, resource=infrastructure)
        return resource

    def get_adapter(self, adapter: Any) -> Any:
        resource = self._get_resource(resource_type=ResourceType.adapter, resource=adapter)
        return resource

    def get_path(self, path_name: str) -> Any:
        return self._get_resource(resource_type=ResourceType.path, resource=path_name)



    def _get_resource(self, resource_type: ResourceType, resource) -> Any:
        assert (resource_type, resource) in self.factory, f"{resource_type.value} for {resource} must be injected first"
        return self.factory[(resource_type, resource)](self)

    @classmethod
    def create_empty(cls) -> 'Dependencies':
        return Dependencies()

