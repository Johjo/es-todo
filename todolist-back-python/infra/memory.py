from uuid import UUID

from expression import Option, Nothing, Some

from hexagon.todolist.aggregate import TodolistSnapshot, TaskSnapshot


class Memory:
    def __init__(self) -> None:
        self.all_todolist: dict[str, TodolistSnapshot] = {}

    def by(self, todolist_name: str) -> Option[TodolistSnapshot]:
        if todolist_name not in self.all_todolist:
            return Nothing
        return Some(self.all_todolist[todolist_name])

    def save(self, todolist: TodolistSnapshot):
        self.all_todolist[todolist.name] = todolist

    def task_by(self, todolist_name: str, task_key: UUID) -> TaskSnapshot:
        all_tasks = self.all_todolist[todolist_name].tasks
        task = [task for task in all_tasks if task.key == task_key][0]
        return task

    def all_todolist_by_name(self) -> list[str]:
        return [name for name in self.all_todolist.keys()]

    def all_tasks(self, todolist_name: str) -> list[TaskSnapshot]:
        if todolist_name not in self.all_todolist:
            return []
        return list(self.all_todolist[todolist_name].tasks)


