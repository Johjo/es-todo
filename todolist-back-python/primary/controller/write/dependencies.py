from abc import ABC
from collections import namedtuple
from dataclasses import dataclass, replace, field
from typing import Any

from hexagon.todolist.aggregate import TodolistSetPort
from hexagon.todolist.write.create_todolist import TodolistCreate
from hexagon.todolist.write.open_task import OpenTask


@dataclass(frozen=True)
class Dependencies(ABC):
    use_case_factory: dict[Any, Any] = field(default_factory=dict)
    adapter_factory: Any = None

    def get_adapter(self, port) -> Any:
        assert self.adapter_factory, f"adapter for {port} must be injected first"
        return self.adapter_factory(self)

    def feed_use_case(self, use_case: Any, use_case_factory: Any) -> 'Dependencies':
        return replace(self, use_case_factory={**self.use_case_factory, use_case: use_case_factory})


    def feed_adapter(self, port: Any, adapter_factory: Any) -> 'Dependencies':
        return replace(self, adapter_factory=adapter_factory)

    def get_use_case(self, use_case: Any) -> Any:
        assert use_case in self.use_case_factory, f"use_case for {use_case} must be injected first"
        return self.use_case_factory[use_case](self)

    @classmethod
    def create_empty(cls) -> 'Dependencies':
        return Dependencies()


def inject_use_cases(dependencies: Dependencies) -> Dependencies:
    factories = {
        TodolistCreate: todolist_create_factory,
        OpenTask : open_task_use_case_factory,
    }

    for use_case, factory in factories.items():
        dependencies = dependencies.feed_use_case(use_case, factory)
    return dependencies


def todolist_create_factory(dependencies: Dependencies):
    return TodolistCreate(dependencies.get_adapter(TodolistSetPort))

def open_task_use_case_factory(dependencies: Dependencies):
    return OpenTask(dependencies.get_adapter(TodolistSetPort))
