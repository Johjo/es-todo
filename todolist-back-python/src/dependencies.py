from dataclasses import dataclass, field, replace
from enum import Enum
from math import factorial
from typing import Any

class ResourceType(str, Enum):
    use_case = "use_case"
    adapter = "adapter"
    path = "path"
    infrastructure = "infrastructure"
    data = "data"


@dataclass(frozen=True)
class Dependencies:
    factory: dict[Any, Any] = field(default_factory=dict)

    def feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return self._feed(resource_type=resource_type, resource=resource, factory=factory)

    def feed_use_case(self, use_case: Any, use_case_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.use_case, resource=use_case, factory=use_case_factory)

    def feed_adapter(self, port: Any, adapter_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.adapter, resource=port, factory=adapter_factory)

    def feed_path(self, path: str, factory) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.path, resource=path, factory=factory)

    def feed_infrastructure(self, infrastructure: Any, factory):
        return self._feed(resource_type=ResourceType.infrastructure, resource=infrastructure, factory=factory)

    def feed_data(self, data_name: str, value: Any):
        return self._feed(resource_type=ResourceType.data, resource=data_name, factory=lambda _: value)

    def _feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return replace(self, factory={**self.factory, (resource_type, resource): factory})

    def get_use_case(self, use_case: Any) -> Any:
        return self._get_resource(resource_type=ResourceType.use_case, resource=use_case)

    def get_query(self, query: Any) -> Any:
        return self.get_use_case(query)

    def get_infrastructure(self, infrastructure) -> Any:
        resource = self._get_resource(resource_type=ResourceType.infrastructure, resource=infrastructure)
        return resource

    def get_adapter(self, adapter: Any) -> Any:
        resource = self._get_resource(resource_type=ResourceType.adapter, resource=adapter)
        return resource

    def get_path(self, path_name: str) -> Any:
        return self._get_resource(resource_type=ResourceType.path, resource=path_name)

    def get_data(self, data_name) -> Any:
        return self._get_resource(resource_type=ResourceType.data, resource=data_name)

    def _get_resource(self, resource_type: ResourceType, resource) -> Any:
        if (resource_type, resource) not in self.factory:
            raise Exception(f"{resource_type.value} for {resource} must be injected first")
        return self.factory[(resource_type, resource)](self)

    @classmethod
    def create_empty(cls) -> 'Dependencies':
        return Dependencies()

    def describe(self) -> str:
        def format_item(key):
            (resource_type, resource_name) = key
            resource = self._get_resource(resource_type, resource_name)
            return f"{resource_name}: {resource_type} => {str(type(resource))}"

        return "\n".join([format_item(key) for key in self.factory.keys()])
