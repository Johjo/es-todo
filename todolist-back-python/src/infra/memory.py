from uuid import UUID

from expression import Option, Nothing, Some

from src.hexagon.todolist.aggregate import TodolistSnapshot, TaskSnapshot


class Memory:
    def __init__(self) -> None:
        self.all_todolist: dict[tuple[str, str], TodolistSnapshot] = {}

    def by(self, user_key: str, todolist_name: str) -> Option[TodolistSnapshot]:
        print(self.all_todolist)
        print(user_key, todolist_name)
        if (user_key, todolist_name) not in self.all_todolist:
            return Nothing
        return Some(self.all_todolist[(user_key, todolist_name)])

    def save(self, user_key: str, todolist: TodolistSnapshot):
        self.all_todolist[(user_key, todolist.name)] = todolist

    def task_by(self, user_key: str, todolist_name: str, task_key: UUID) -> TaskSnapshot:
        all_tasks = self.all_todolist[(user_key, todolist_name)].tasks
        task = [task for task in all_tasks if task.key == task_key][0]
        return task

    def all_todolist_by_name(self, user_key: str) -> list[str]:
        return [name for (todolist_user_key, name) in self.all_todolist.keys() if todolist_user_key == user_key]

    def all_tasks(self, user_key: str, todolist_name: str) -> list[TaskSnapshot]:
        if (user_key, todolist_name) not in self.all_todolist:
            return []
        return list(self.all_todolist[(user_key, todolist_name)].tasks)

    def __repr__(self):
        return str(self.all_todolist)


