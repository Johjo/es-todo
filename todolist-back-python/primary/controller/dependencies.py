from functools import reduce
from typing import Any

from dependencies import Dependencies
from hexagon.fvp.read.which_task import WhichTaskQuery
from hexagon.fvp.write.cancel_priority import CancelPriority
from hexagon.fvp.write.choose_and_ignore_task import ChooseAndIgnoreTaskFvp
from hexagon.fvp.write.reset_fvp_session import ResetFvpSession
from hexagon.todolist.write.close_task import CloseTask
from hexagon.todolist.write.create_todolist import TodolistCreate
from hexagon.todolist.write.import_many_task import ImportManyTask
from hexagon.todolist.write.open_task import OpenTaskUseCase
from hexagon.todolist.write.reword_task import RewordTask


def inject_use_cases(dependencies: Dependencies) -> Dependencies:
    use_cases_with_factory: list[Any] = [ResetFvpSession, CancelPriority, ChooseAndIgnoreTaskFvp, TodolistCreate,
                                         OpenTaskUseCase, CloseTask, RewordTask, ImportManyTask, WhichTaskQuery]

    def feed_use_case(dep: Dependencies, use_case: Any) -> Dependencies:
        return dep.feed_use_case(use_case=use_case, use_case_factory=use_case.factory)

    return reduce(feed_use_case, use_cases_with_factory, dependencies)


