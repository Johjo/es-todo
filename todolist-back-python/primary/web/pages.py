import urllib
from dataclasses import dataclass
from uuid import UUID

from bottle import template, Bottle, view, response, request, redirect  # type: ignore

from hexagon.fvp.aggregate import NothingToDo, DoTheTask, ChooseTheTask
from hexagon.fvp.read.which_task import TaskFilter
from hexagon.shared.type import TaskKey, TodolistName
from primary.controller.read.final_version_perfected import FinalVersionPerfectedReadController
from primary.controller.read.todolist import TodolistReadController, Task
from dependencies import Dependencies
from primary.controller.write.todolist import TodolistWriteController

bottle_app = Bottle()


@dataclass
class BottleConfig:
    dependencies: Dependencies


bottle_config = BottleConfig(dependencies=Dependencies.create_empty())


@bottle_app.route("/")
@view("index")
def index():
    controller = TodolistReadController(bottle_config.dependencies)
    return {"todolist_name_set": controller.all_todolist_by_name()}


@bottle_app.post("/todo")
def create_todolist():
    todolist_name = get_string_from_request_post("name")
    TodolistWriteController(bottle_config.dependencies).create_todolist(todolist_name)
    redirect(f"/todo/{todolist_name}")


@bottle_app.route("/todo/<todolist_name>/import")
@view("import")
def show_import(todolist_name):
    return {"todolist_name": todolist_name,
            "query_string": "xxxxquery_string",
            "number_of_items": "xxxx_number_of_items",
            "counts_by_context": {"xxxx1": "xxxx2"},
            "urlencode": urllib.parse.quote,
            }


@bottle_app.post("/todo/<todolist_name>/import")
def import_task(todolist_name):
    markdown_import = get_string_from_request_post("markdown_import")
    TodolistWriteController(bottle_config.dependencies).import_many_tasks_from_markdown(todolist_name, markdown_import)
    return redirect(f"/todo/{todolist_name}")


@bottle_app.route("/todo/<todolist_name>")
def show_todolist(todolist_name):
    task_filter = TaskFilter(todolist_name=todolist_name)  # todo mutate this

    which_task = FinalVersionPerfectedReadController(bottle_config.dependencies).which_task(task_filter=task_filter)
    match which_task:
        case NothingToDo():
            return show_todolist_when_no_task(todolist_name=todolist_name)

        case DoTheTask():
            return show_todolist_when_one_task(todolist_name=todolist_name, do_the_task=which_task)

        case ChooseTheTask():
            return show_todolist_when_two_tasks(todolist_name=todolist_name, choose_the_task=which_task)


def show_todolist_when_no_task(todolist_name):
    return template("nothing", {
        "todolist_name": todolist_name,
        "query_string": "xxxx_query_string",
        "number_of_items": "xxxx_number_of_items",
        "counts_by_context": {"xxxx1": "xxxx2"},
        "urlencode": urllib.parse.quote,
    })


def show_todolist_when_one_task(todolist_name: str, do_the_task: DoTheTask):
    return template("do_the_task", {
        "task_name": do_the_task.name,
        "task_id": do_the_task.id,
        "todolist_name": todolist_name,
        "query_string": "xxxx_query_string",
        "number_of_items": "xxxx_number_of_items",
        "counts_by_context": {"xxxx1": "xxxx2"},
        "urlencode": urllib.parse.quote}, )


def show_todolist_when_two_tasks(todolist_name: str, choose_the_task: ChooseTheTask):
    return template("choose_the_task", {
        "name_1": choose_the_task.name_1,
        "index_1": choose_the_task.id_1,
        "todolist_name": todolist_name,
        "index_2": choose_the_task.id_2,
        "name_2": choose_the_task.name_2,
        "query_string": "xxxx_query_string",
        "number_of_items": "xxxx_number_of_items",
        "counts_by_context": {"xxxx1": "xxxx2"},
        "urlencode": urllib.parse.quote,
    })


@bottle_app.post("/todo/<todolist_name>/item")
def open_task(todolist_name: str):
    task_name = get_string_from_request_post("task_name")
    TodolistWriteController(bottle_config.dependencies).open_task(todolist_name, task_name)
    return redirect(f"/todo/{todolist_name}")


@bottle_app.post("/todo/<todolist_name>/item/<task_key>/close")
def close_task(todolist_name: str, task_key: str):
    TodolistWriteController(bottle_config.dependencies).close_task(todolist_name, task_key=TaskKey(UUID(task_key)))
    return redirect(f"/todo/{todolist_name}")


@bottle_app.post("/todo/<todolist_name>/item/<task_key>/reword")
def reword_task(todolist_name: str, task_key: str):
    controller = TodolistWriteController(bottle_config.dependencies)
    controller.reword_task(todolist_name=todolist_name,
                           task_key=TaskKey(UUID(task_key)),
                           new_name=get_string_from_request_post("new_name"))

    return redirect(f"/todo/{todolist_name}")


@bottle_app.get("/todo/<todolist_name>/item/<task_key>/reword")
def display_reword_task(todolist_name: str, task_key: str):
    task: Task = TodolistReadController(bottle_config.dependencies).task_by(todolist_name=todolist_name,
                                                                            task_key=TaskKey(UUID(task_key)))
    return template("reword", {
        "todolist_name": todolist_name,
        "task_id": task_key,
        "query_string": "xxxx_query_string",
        "task_name": task.name})


@bottle_app.get("/todo/<todolist_name>/export")
@view("export")
def display_export_as_markdown(todolist_name: str):
    return {"markdown_export": TodolistReadController(dependencies=bottle_config.dependencies).to_markdown(todolist_name=todolist_name),
            "todolist_name": todolist_name,
            "query_string": "xxx",
            "number_of_items": "xxx",
            "counts_by_context": {"xxx": "xxx"},
            "urlencode": urllib.parse.quote,
            }

@bottle_app.post('/todo/<todolist_name>/item/choose/<chosen_task>/ignore/<ignored_task>')
def choose_and_ignore_task(todolist_name, chosen_task, ignored_task):
    TodolistWriteController(bottle_config.dependencies).choose_and_ignore_task(chosen_task=chosen_task,
                                                                               ignored_task=ignored_task)
    return redirect(f"/todo/{todolist_name}")

@bottle_app.post('/todo/<todolist_name>/item/<task_key>/cancel_priority')
def cancel_priority(todolist_name, task_key):
    controller = TodolistWriteController(bottle_config.dependencies)
    controller.cancel_priority(task_key=TaskKey(UUID(task_key)))
    return redirect(f"/todo/{todolist_name}")

def get_string_from_request_post(field_name):
    return request.forms.getunicode(field_name)
