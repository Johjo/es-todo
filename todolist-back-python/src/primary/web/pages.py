from dataclasses import dataclass
from datetime import datetime, date
from pathlib import Path
from typing import Any
from urllib.parse import quote as urlencode
from uuid import UUID

from bottle import template, Bottle, view, request, redirect, static_file, auth_basic  # type: ignore
from bottle_utils import form  # type: ignore

from src.dependencies import Dependencies
from src.hexagon.fvp.aggregate import NothingToDo, DoTheTask, ChooseTheTask
from src.hexagon.fvp.read.which_task import WhichTaskFilter
from src.hexagon.shared.type import TaskKey, TodolistName, TaskExecutionDate, TaskName
from src.primary.controller.read.final_version_perfected import FinalVersionPerfectedReadController
from src.primary.controller.read.todolist import TodolistReadController, TaskPresentation
from src.primary.controller.write.todolist import TodolistWriteController
from src.shared.const import USER_KEY

bottle_app = Bottle()


@dataclass
class BottleConfig:
    dependencies: Dependencies


bottle_config = BottleConfig(dependencies=Dependencies.create_empty())


def is_authenticated_user(user, password):
    return user == password


@bottle_app.route("/todo")
@bottle_app.route("/")
@auth_basic(is_authenticated_user)
@view("index")
def index():
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    controller = TodolistReadController(bottle_config.dependencies)
    return {"todolist_name_set": controller.all_todolist_by_name()}


@bottle_app.post("/todo")
@auth_basic(is_authenticated_user)
def create_todolist():
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    todolist_name = get_string_from_request_post("name")
    TodolistWriteController(bottle_config.dependencies).create_todolist(TodolistName(todolist_name))
    return redirect_to_todolist(todolist_name)


@bottle_app.route("/todo/<todolist_name>/import")
@view("import")
@auth_basic(is_authenticated_user)
def show_import(todolist_name):
    return base_value(todolist_name)


@bottle_app.route("/todo/<todolist_name>/calendar")
@view("calendar")
@auth_basic(is_authenticated_user)
def show_calendar(todolist_name):
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    tasks = TodolistReadController(bottle_config.dependencies).all_tasks_postponed_task(todolist_name)
    return {**base_value(todolist_name),
            "tasks":tasks}


@bottle_app.post("/todo/<todolist_name>/import")
@auth_basic(is_authenticated_user)
def import_task(todolist_name):
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    markdown_import = get_string_from_request_post("markdown_import")
    TodolistWriteController(bottle_config.dependencies).import_many_tasks_from_markdown(todolist_name, markdown_import)
    return redirect_to_todolist(todolist_name)


@bottle_app.route("/todo/<todolist_name>")
@auth_basic(is_authenticated_user)
def show_todolist(todolist_name: str) -> str:
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    task_filter = WhichTaskFilter(todolist_name=TodolistName(todolist_name),
                                  include_context=tuple(request.query.getall('include_context')),
                                  exclude_context=tuple(request.query.getall('exclude_context')),
                                  reference_date=date(2050, 10, 17))

    controller = FinalVersionPerfectedReadController(bottle_config.dependencies)
    which_task = controller.which_task(
        todolist_name=todolist_name,
        include_context=tuple(request.query.getall('include_context')),
        exclude_context=tuple(request.query.getall('exclude_context')),
        task_filter=task_filter)

    match which_task:
        case NothingToDo():
            return show_todolist_when_no_task(todolist_name=todolist_name)

        case DoTheTask():
            return show_todolist_when_one_task(todolist_name=todolist_name, do_the_task=which_task)

        case ChooseTheTask():
            return show_todolist_when_two_tasks(todolist_name=todolist_name, choose_the_task=which_task)


def show_todolist_when_no_task(todolist_name):
    return template("nothing", base_value(todolist_name))


def show_todolist_when_one_task(todolist_name: str, do_the_task: DoTheTask):
    task = TodolistReadController(bottle_config.dependencies).task_by(todolist_name, do_the_task.key)
    return template("do_the_task",
                    {**base_value(todolist_name),
                     "task_name": task.name,
                     "task_id": task.key,
                     "task": task,
                     }, )


def show_todolist_when_two_tasks(todolist_name: str, choose_the_task: ChooseTheTask):
    task_1 = TodolistReadController(bottle_config.dependencies).task_by(todolist_name, choose_the_task.main_task_key)
    task_2 = TodolistReadController(bottle_config.dependencies).task_by(todolist_name, choose_the_task.secondary_task_key)

    return template("choose_the_task",
                    {**base_value(todolist_name),
                     "todolist_name": todolist_name,
                     "task_1": task_1,
                     "task_2": task_2,
                     })


@bottle_app.post("/todo/<todolist_name>/item")
@auth_basic(is_authenticated_user)
def open_task(todolist_name: str):
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    task_name = get_string_from_request_post("task_name")
    TodolistWriteController(bottle_config.dependencies).open_task(TodolistName(todolist_name), TaskName(task_name))
    return redirect_to_todolist(todolist_name)


def redirect_to_todolist(todolist_name) -> str:
    return redirect(f"/todo/{todolist_name}?{request.query_string}")


@bottle_app.post("/todo/<todolist_name>/item/<task_key>/close")
@auth_basic(is_authenticated_user)
def close_task(task_key: str, todolist_name: str):
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    TodolistWriteController(bottle_config.dependencies).close_task(TodolistName(todolist_name), task_key=TaskKey(UUID(task_key)))
    return redirect_to_todolist(todolist_name)


@bottle_app.post("/todo/<todolist_name>/item/<task_key>/reword")
@auth_basic(is_authenticated_user)
def reword_task(todolist_name: str, task_key: str) -> str:
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    controller = TodolistWriteController(bottle_config.dependencies)
    controller.reword_task(todolist_name=TodolistName(todolist_name),
                           task_key=TaskKey(UUID(task_key)),
                           new_name=TaskName(get_string_from_request_post("new_name")))

    return redirect_to_todolist(todolist_name)


@bottle_app.get("/todo/<todolist_name>/item/<task_key>/reword")
@auth_basic(is_authenticated_user)
def display_reword_task(todolist_name: str, task_key: str) -> str:
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    controller = TodolistReadController(bottle_config.dependencies)

    task: TaskPresentation = controller.task_by(todolist_name=todolist_name,
                                                task_key=TaskKey(UUID(task_key)))
    return template("reword", {
        "todolist_name": todolist_name,
        "task_id": task_key,
        "query_string": request.query_string,
        "task_name": task.name})


class PostponeForm(form.Form):
    execution_date = form.DateField(label="Date d'execution")


@bottle_app.get("/todo/<todolist_name>/item/<task_key>/postpone")
@auth_basic(is_authenticated_user)
def display_postpone_task(todolist_name: str, task_key: str) -> str:
    postpone_form = PostponeForm()
    return display_form_postpone_task(todolist_name, task_key, postpone_form)


def display_form_postpone_task(todolist_name, task_key, postpone_form) -> str:
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    controller = TodolistReadController(bottle_config.dependencies)
    task = controller.task_by(todolist_name=todolist_name, task_key=TaskKey(UUID(task_key)))
    return template("postpone", {
        "todolist_name": todolist_name,
        "task_id": task_key,
        "query_string": request.query_string,
        "task": task,
        "form": postpone_form,
    })


@bottle_app.post("/todo/<todolist_name>/item/<task_key>/postpone")
@auth_basic(is_authenticated_user)
def postpone(todolist_name: str, task_key: str) -> str:
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    postpone_form = PostponeForm(request.forms)
    if postpone_form.is_valid():
        controller = TodolistWriteController(bottle_config.dependencies)
        date_as_str = request.forms.getunicode("execution_date")
        controller.postpone_task(
            name=TodolistName(todolist_name),
            key=TaskKey(UUID(task_key)),
            execution_date=TaskExecutionDate(datetime.strptime(date_as_str, "%Y-%m-%d").date()))
        return redirect_to_todolist(todolist_name)
    return display_form_postpone_task(todolist_name, task_key, postpone_form)


@bottle_app.post("/todo/<todolist_name>/item/<task_key>/tomorrow")
@auth_basic(is_authenticated_user)
def postpone_to_tomorrow(todolist_name: str, task_key: str) -> str:
    controller = TodolistWriteController(bottle_config.dependencies)
    controller.postpone_task_to_tomorrow(name=TodolistName(todolist_name), key=TaskKey(UUID(task_key)))
    return redirect_to_todolist(todolist_name)

    # bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    # postpone_form = PostponeForm(request.forms)
    # if postpone_form.is_valid():
    #     controller = TodolistWriteController(bottle_config.dependencies)
    #     date_as_str = request.forms.getunicode("execution_date")
    #     controller.postpone_task(
    #         name=TodolistName(todolist_name),
    #         key=TaskKey(UUID(task_key)),
    #         execution_date=TaskExecutionDate(datetime.strptime(date_as_str, "%Y-%m-%d").date()))
    #     return redirect_to_todolist(todolist_name)
    # return display_form_postpone_task(todolist_name, task_key, postpone_form)


@bottle_app.get("/todo/<todolist_name>/export")
@view("export")
@auth_basic(is_authenticated_user)
def display_export_as_markdown(todolist_name: str) -> dict:
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    markdown = TodolistReadController(dependencies=bottle_config.dependencies).to_markdown(todolist_name=todolist_name)

    return {**base_value(todolist_name), "markdown_export": markdown, }


def authenticate(dependencies: Dependencies, user: str) -> Dependencies:
    return dependencies.feed_data(data_name=USER_KEY, value=user)


@bottle_app.post('/todo/<todolist_name>/item/choose/<chosen_task>/ignore/<ignored_task>')
@auth_basic(is_authenticated_user)
def choose_and_ignore_task(todolist_name: str, chosen_task:  str, ignored_task:str):
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    TodolistWriteController(bottle_config.dependencies).choose_and_ignore_task(chosen_task=TaskKey(UUID(chosen_task)),
                                                                               ignored_task=TaskKey(UUID(ignored_task)))
    return redirect_to_todolist(todolist_name)


@bottle_app.post('/todo/<todolist_name>/item/<task_key>/cancel_priority')
@auth_basic(is_authenticated_user)
def cancel_priority(todolist_name, task_key):
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    controller = TodolistWriteController(bottle_config.dependencies)
    controller.cancel_priority(task_key=TaskKey(UUID(task_key)))
    return redirect_to_todolist(todolist_name)


@bottle_app.post('/todo/<todolist_name>/reset')
@auth_basic(is_authenticated_user)
def reset_all_priorities(todolist_name):
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    controller = TodolistWriteController(bottle_config.dependencies)
    controller.reset_all_priorities()
    return redirect_to_todolist(todolist_name)


@bottle_app.get('/static/<filename:path>')
def static(filename: str) -> Any:
    root : str = bottle_config.dependencies.get_path("static_path")
    return static_file(filename, root=Path(root).absolute())





def get_string_from_request_post(field_name: str) -> str:
    return str(request.forms.getunicode(field_name))


def base_value(todolist_name: str) -> dict[str, Any]:
    bottle_config.dependencies = authenticate(dependencies=bottle_config.dependencies, user=request.auth[0])
    return {
        "todolist_name": todolist_name,
        "query_string": request.query_string,
        "number_of_items": "xxxx_number_of_items",
        "counts_by_context": TodolistReadController(bottle_config.dependencies).counts_by_context(todolist_name),
        "included_context": request.query.getall("include_context"),
        "excluded_context": request.query.getall("exclude_context"),
        "urlencode": urlencode,
    }
