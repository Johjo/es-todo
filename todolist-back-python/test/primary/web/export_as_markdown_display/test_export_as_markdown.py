from approvaltests import verify  # type: ignore
from approvaltests.reporters import PythonNativeReporter
from webtest import TestApp  # type: ignore

from dependencies import Dependencies
from infra.memory import Memory
from primary.web.pages import bottle_config
from test.fixture import TodolistFaker


def test_display_export_as_markdown(memory: Memory, test_dependencies: Dependencies, app: TestApp, fake: TodolistFaker):
    bottle_config.dependencies = test_dependencies

    todolist = fake.a_todolist().having(name="todolist").having(tasks=[fake.a_task().having(name="buy milk"), fake.a_task().having(name="buy water")])
    memory.save(todolist.to_snapshot())

    response = app.get(f'/todo/{todolist.name}/export')

    assert response.status == '200 OK'
    verify(str(response.body).replace("\\r\\n", "\r\n"), reporter=PythonNativeReporter())
