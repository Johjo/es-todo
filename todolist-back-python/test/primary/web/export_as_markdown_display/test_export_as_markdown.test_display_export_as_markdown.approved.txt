b'<html>

<head>

    <meta charset="UTF-8">

    <title>Todo</title>

</head>

<body>

<div><a href="/todo/todolist/calendar?">Calendrier</a></div>



<p>

<textarea rows="10" cols="80" readonly>- [ ] buy milk\n- [ ] buy water</textarea>



</p>



<div>

Ajouter une t\xc3\xa2che

<form action="/todo/todolist/item?" method="post">

    <input type="text" name="task_name">

    <input type="submit" value="Add item">

</form>

</div>



Red\xc3\xa9marrer l\'algo

<form action="/todo/todolist/reset?" method="post">

    <input type="submit" value="Reset">

</form>



Il y a xxxx_number_of_items t\xc3\xa2ches en attente.





<h2> Contexte </h2>

<form action="/todo/todolist" method="get">

<ul>

<li> <a href="/todo/todolist">Supprimer les filtres</a></li>

</ul>

    <input type="submit" value="Filtrer">

</form>



<ul>

<li><a href="/todo/todolist/export">Exporter les t\xc3\xa2ches</a></li>

<li><a href="/todo/todolist/import">Importer des t\xc3\xa2ches</a></li>

</ul>







</body>

</html>'
